//! Error types definitions.
//!
//! All usage errors will pass throught the Error enum, a lot of them in the Error::Custom.

use std::{
    fmt::{self, Display},
    path::{Path, PathBuf},
};

use crate::utils::colors::*;

/// All errors that can be generated by `ouch`
#[derive(Debug, PartialEq)]
pub enum Error {
    /// Extension found is not supported and known to ouch
    UnknownExtensionError(String),
    /// TO BE REMOVED
    MissingExtensionError(PathBuf),
    /// Not every IoError, some of them get filtered by `From<io::Error>` into other variants
    IoError {
        /// TODO
        reason: String,
    },
    /// Detected from io::Error if .kind() is io::ErrorKind::NotFound
    FileNotFound(PathBuf),
    /// TO BE REMOVED
    AlreadyExists,
    /// TO BE REMOVED
    InvalidZipArchive(&'static str),
    /// Detected from io::Error if .kind() is io::ErrorKind::PermissionDenied
    PermissionDenied {
        /// TODO
        error_title: String,
    },
    /// TO BE REMOVED
    UnsupportedZipArchive(&'static str),
    /// TO BE REMOVED
    InternalError,
    /// TO BE REMOVED
    CompressingRootFolder,
    /// TO BE REMOVED
    MissingArgumentsForCompression,
    /// TO BE REMOVED
    MissingArgumentsForDecompression,
    /// TO BE REMOVED
    CompressionTypo,
    /// Specialized walkdir's io::Error wrapper with additional information on the error
    WalkdirError {
        /// TODO
        reason: String,
    },
    /// Custom and unique errors are reported in this variant
    Custom {
        /// TODO
        reason: FinalError,
    },
}

/// Alias to std's Result with ouch's Error
pub type Result<T> = std::result::Result<T, Error>;

/// Pretty final error message for end users, crashing the program after display.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct FinalError {
    /// Should be made of just one line, appears after the "[ERROR]" part
    title: String,
    /// Shown as a unnumbered list in yellow
    details: Vec<String>,
    /// Shown as green at the end to give hints on how to work around this error, if it's fixable
    hints: Vec<String>,
}

impl Display for FinalError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Title
        writeln!(f, "{}[ERROR]{} {}", *RED, *RESET, self.title)?;

        // Details
        for detail in &self.details {
            writeln!(f, " {}-{} {}", *WHITE, *YELLOW, detail)?;
        }

        // Hints
        if !self.hints.is_empty() {
            // Separate by one blank line.
            writeln!(f)?;
            for hint in &self.hints {
                writeln!(f, "{}hint:{} {}", *GREEN, *RESET, hint)?;
            }
        }

        write!(f, "{}", *RESET)
    }
}

impl FinalError {
    /// Only constructor
    pub fn with_title(title: impl ToString) -> Self {
        Self { title: title.to_string(), details: vec![], hints: vec![] }
    }

    /// Add one detail line, can have multiple
    pub fn detail(mut self, detail: impl ToString) -> Self {
        self.details.push(detail.to_string());
        self
    }

    /// Add one hint line, can have multiple
    pub fn hint(mut self, hint: impl ToString) -> Self {
        self.hints.push(hint.to_string());
        self
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let err = match self {
            Error::MissingExtensionError(filename) => {
                FinalError::with_title(format!("Cannot compress to {:?}", filename))
                    .detail("Ouch could not detect the compression format")
                    .hint("Use a supported format extension, like '.zip' or '.tar.gz'")
                    .hint("Check https://github.com/ouch-org/ouch for a full list of supported formats")
            }
            Error::WalkdirError { reason } => FinalError::with_title(reason),
            Error::FileNotFound(file) => {
                if file == Path::new("") {
                    FinalError::with_title("file not found!")
                } else {
                    FinalError::with_title(format!("file {:?} not found!", file))
                }
            }
            Error::CompressingRootFolder => {
                FinalError::with_title("It seems you're trying to compress the root folder.")
                    .detail("This is unadvisable since ouch does compressions in-memory.")
                    .hint("Use a more appropriate tool for this, such as rsync.")
            }
            Error::MissingArgumentsForCompression => {
                FinalError::with_title("Could not compress")
                    .detail("The compress command requires at least 2 arguments")
                    .hint("You must provide:")
                    .hint("  - At least one input argument.")
                    .hint("  - The output argument.")
                    .hint("")
                    .hint("Example: `ouch compress image.png img.zip`")
            }
            Error::MissingArgumentsForDecompression => {
                FinalError::with_title("Could not decompress")
                    .detail("The compress command requires at least one argument")
                    .hint("You must provide:")
                    .hint("  - At least one input argument.")
                    .hint("")
                    .hint("Example: `ouch decompress imgs.tar.gz`")
            }
            Error::InternalError => {
                FinalError::with_title("InternalError :(")
                    .detail("This should not have happened")
                    .detail("It's probably our fault")
                    .detail("Please help us improve by reporting the issue at:")
                    .detail(format!("    {}https://github.com/ouch-org/ouch/issues ", *CYAN))
            }
            Error::IoError { reason } => FinalError::with_title(reason),
            Error::CompressionTypo => {
                FinalError::with_title("Possible typo detected")
                    .hint(format!("Did you mean '{}ouch compress{}'?", *MAGENTA, *RESET))
            }
            Error::UnknownExtensionError(_) => todo!(),
            Error::AlreadyExists => todo!(),
            Error::InvalidZipArchive(_) => todo!(),
            Error::PermissionDenied { error_title } => FinalError::with_title(error_title).detail("Permission denied"),
            Error::UnsupportedZipArchive(_) => todo!(),
            Error::Custom { reason } => reason.clone(),
        };

        write!(f, "{}", err)
    }
}

impl From<std::io::Error> for Error {
    fn from(err: std::io::Error) -> Self {
        match err.kind() {
            std::io::ErrorKind::NotFound => todo!(),
            std::io::ErrorKind::PermissionDenied => Self::PermissionDenied { error_title: err.to_string() },
            std::io::ErrorKind::AlreadyExists => Self::AlreadyExists,
            _other => Self::IoError { reason: err.to_string() },
        }
    }
}

impl From<zip::result::ZipError> for Error {
    fn from(err: zip::result::ZipError) -> Self {
        use zip::result::ZipError::*;
        match err {
            Io(io_err) => Self::from(io_err),
            InvalidArchive(filename) => Self::InvalidZipArchive(filename),
            FileNotFound => Self::FileNotFound("".into()),
            UnsupportedArchive(filename) => Self::UnsupportedZipArchive(filename),
        }
    }
}

impl From<walkdir::Error> for Error {
    fn from(err: walkdir::Error) -> Self {
        Self::WalkdirError { reason: err.to_string() }
    }
}

impl From<FinalError> for Error {
    fn from(err: FinalError) -> Self {
        Self::Custom { reason: err }
    }
}
